src/util/tmpl/fd_heap.c

```diff
@@ -0,0 +1,647 @@
+   persisted beyond the lifetime of creating process, used concurrently
+   in many common operations, used inter-process, relocated in memory,
+   naively serialized/deserialized, moved between hosts, supports index
+   compresson for cache and memory bandwidth efficiency, etc.
+
+   Typical usage:
+
+     struct myele {
+
+       ... Each field below can be located arbitrarily in the struct
+
+       ulong left;  // Technically "HEAP_IDX_T HEAP_LEFT;" (default is ulong left), similarly for right.  These are managed by the
+       ulong right; // heap when a myele is in the heap.
+
+       ... Note that other kinds of objects can use these fields for
+       ... their metadata needs to keep element metadata / cache
+       ... footprint overheads minimal.  The only restriction is that
+       ... they cannot concurrently use the same field.
+
+       ... Note that fields could be in an anonymous union and/or made
+       ... into narrow bit fields if useful for additional layout,
+       ... memory, bandwidth and cache efficiency.
+
+       ... Arbitrary application fields mixed in here.  Power-of-2
+       ... element sizes have good cache and indexing Feng Shui.
+
+       char key[ KEY_MAX ]; // For demonstration purposes
+
+     };
+
+     typedef struct myele myele_t;
+
+     #define   HEAP_NAME      myheap
+     #define   HEAP_T         myele_t
+     #define   HEAP_LT(e0,e1) (strcmp( e0->key, e1->key )<0)
+     #include "fd_heap.c"
+
+   will declare the following APIs as a header-only style library in the
+   compilation unit:
+
+     int myheap_lt( myele_t const * e0, myele_t const * e1 ); // Provides HEAP_LT
+
+     // myheap_idx_null returns the element index used to represent
+     // NULL, infinite lifetime.  myheap_ele_null returns NULL, infinite
+     // lifetime, for completeness, myheap_ele_null_const is a
+     // const-correct version, also for completeness.
+
+     ulong           myheap_idx_null      ( void );
+     myele_t *       myheap_ele_null      ( void );
+     myele_t const * myheap_ele_null_const( void );
+
+     // myheap_{idx,ele}_is_null returns i==myheap_idx_null() / !e
+
+     int myheap_idx_is_null( ulong           i );
+     int myheap_ele_is_null( myele_t const * e );
+
+     // myheap_idx returns e's index.  Assumes e in ele or is NULL.
+     // Return will be in [0,ele_max) or myheap_idx_null().  Lifetime is
+     // the element storage lifetime.  myheap_idx_fast is the same but
+     // assumes e is not NULL.  myheap_ele returns a pointer in the
+     // caller's address space to element idx.  Assumes idx is in
+     // [0,ele_max) or is myheap_idx_null().  Return pointer lifetime is
+     // ele's local lifetime.  myheap_ele_fast is the same but assumes
+     // idx is not myheap_idx_null().  myheap_ele[_fast]_const is a
+     // const correct version.
+
+     ulong           myheap_idx           ( myele_t const * e, myele_t const * ele );
```

Comment:
What is the ele argument here? Some sentinel element? It's not clear what your indexes are relative to.